crypto/evp/m_md5_sha1.c

```diff
@@ -42,82 +48,84 @@ static int update(EVP_MD_CTX *ctx, const void *data, size_t count)
 
-    /* SSLv3 client auth handling: see RFC-6101 5.6.8 */
-    if (mslen != 48)
-        return 0;
+        if (!MD5_Update(&mctx->md5, padtmp, sizeof(padtmp)))
+            return 0;
 
-    /* At this point hash contains all handshake messages, update
-     * with master secret and pad_1.
-     */
+        if (!MD5_Final(md5tmp, &mctx->md5))
+            return 0;
 
-    if (update(ctx, ms, mslen) <= 0)
-        return 0;
+        if (!SHA1_Update(&mctx->sha1, padtmp, 40))
+            return 0;
 
-    /* Set padtmp to pad_1 value */
-    memset(padtmp, 0x36, sizeof(padtmp));
+        if (!SHA1_Final(sha1tmp, &mctx->sha1))
+            return 0;
 
-    if (!MD5_Update(&mctx->md5, padtmp, sizeof(padtmp)))
-        return 0;
+        /* Reinitialise context */
+        if (!MD5_Init(&mctx->md5) || !SHA1_Init(&mctx->sha1))
+            return 0;
 
-    if (!MD5_Final(md5tmp, &mctx->md5))
-        return 0;
+        if (update(ctx, mctx->tmpms, SSL3_MASTER_SECRET_SIZE) <= 0)
+            return 0;
 
-    if (!SHA1_Update(&mctx->sha1, padtmp, 40))
-        return 0;
+        /* Set padtmp to pad_2 value */
+        memset(padtmp, 0x5c, sizeof(padtmp));
 
-    if (!SHA1_Final(sha1tmp, &mctx->sha1))
-        return 0;
+        if (!MD5_Update(&mctx->md5, padtmp, sizeof(padtmp)))
+            return 0;
 
-    /* Reinitialise context */
+        if (!MD5_Update(&mctx->md5, md5tmp, sizeof(md5tmp)))
+            return 0;
 
-    if (!init(ctx))
-        return 0;
+        if (!SHA1_Update(&mctx->sha1, padtmp, 40))
+            return 0;
 
-    if (update(ctx, ms, mslen) <= 0)
-        return 0;
+        if (!SHA1_Update(&mctx->sha1, sha1tmp, sizeof(sha1tmp)))
+            return 0;
 
-    /* Set padtmp to pad_2 value */
-    memset(padtmp, 0x5c, sizeof(padtmp));
+        OPENSSL_cleanse(md5tmp, sizeof(md5tmp));
+        OPENSSL_cleanse(sha1tmp, sizeof(sha1tmp));
+        OPENSSL_clear_free(mctx->tmpms, SSL3_MASTER_SECRET_SIZE);
+        mctx->tmpms = NULL;
+    }
 
-    if (!MD5_Update(&mctx->md5, padtmp, sizeof(padtmp)))
+    if (!MD5_Final(md, &mctx->md5))
         return 0;
+    return SHA1_Final(md + MD5_DIGEST_LENGTH, &mctx->sha1);
+}
 
-    if (!MD5_Update(&mctx->md5, md5tmp, sizeof(md5tmp)))
-        return 0;
+static int ctrl(EVP_MD_CTX *ctx, int cmd, int mslen, void *ms)
+{
+    struct md5_sha1_ctx *mctx;
 
-    if (!SHA1_Update(&mctx->sha1, padtmp, 40))
-        return 0;
+    if (cmd != EVP_CTRL_SSL3_MASTER_SECRET)
+        return -2;
 
-    if (!SHA1_Update(&mctx->sha1, sha1tmp, sizeof(sha1tmp)))
+    if (ctx == NULL)
         return 0;
 
-    /* Now when ctx is finalised it will return the SSL v3 hash value */
-
-    OPENSSL_cleanse(md5tmp, sizeof(md5tmp));
-    OPENSSL_cleanse(sha1tmp, sizeof(sha1tmp));
-
-    return 1;
+    mctx = EVP_MD_CTX_md_data(ctx);
 
+    if (mslen != SSL3_MASTER_SECRET_SIZE)
+        return 0;
+    OPENSSL_free(mctx->tmpms);
+    return (mctx->tmpms = OPENSSL_memdup(ms, SSL3_MASTER_SECRET_SIZE)) != NULL;
```

Comment:
Certainly that would be easier to review. I was also considering the overhead of this change.

This would get added to every MD5+SHA1 context which will be used by TLS 1.1, 1.0 for RSA signatures. The additional data will be used in one single case: SSLv3 with client auth which we don't even compile in by default. If we didn't consider that significant I'd no problems adding a fixed master secret buffer with a "used" flag.

However on reflection this change alters the way the ctrl operates and could break things (albeit in a very rare use case) in a stable release. I think instead we should change libssl so it keeps the old ordering for the one special case. I'll do that in a new PR.